### PR TITLE
Expose tensor check from guard for reusing

### PR DESCRIPTION
**Pull request by EikanWang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124926
* #124070
* #124177
* #116368
* __->__ #124836

cc @voznesenskym @penguinwu @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/csrc/dynamo/guards.h [hunk 1]
```diff
@@ -8,3 +9,67 @@ PyObject* torch_c_dynamo_guards_init();
 // not visible there.
 void* convert_to_root_guard_manager(py::object root);
 bool run_root_guard_manager(void* root, PyObject* f_locals);
+
+struct LocalState {
```
Comment (desertfire) [inline torch/csrc/dynamo/guards.h:13]: Should we add namespace to avoid naming collision?

Comment (EikanWang) [reply torch/csrc/dynamo/guards.h:13]: Yes, we should, and the namespace should be `namespace torch::dynamo`. I'm just not sure about the namespace scope. Should we add `convert_to_root_guard_manager` and `run_root_guard_manager` to the namespace as well? @anijain2305

Comment (jansel) [reply torch/csrc/dynamo/guards.h:13]: Let's move everything to `torch::dynamo` scope

Comment (EikanWang) [reply torch/csrc/dynamo/guards.h:13]: Thanks, I will refine it.

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Review by desertfire — APPROVED** [review]

**Pull request closed without merging**